rustworkx-core/src/token_swapper.rs

```diff
@@ -0,0 +1,608 @@
+use std::hash::Hash;
+
+use hashbrown::HashMap;
+use petgraph::stable_graph::{NodeIndex, StableGraph};
+use petgraph::visit::{
+    EdgeCount, GraphBase, IntoEdges, IntoNeighborsDirected, IntoNodeIdentifiers, NodeCount,
+    NodeIndexable, Visitable,
+};
+use petgraph::Directed;
+use petgraph::Direction::{Incoming, Outgoing};
+use rayon_cond::CondIterator;
+
+use crate::connectivity::find_cycle;
+use crate::dictmap::*;
+use crate::shortest_path::dijkstra;
+use crate::traversal::dfs_edges;
+
+type Swap = (NodeIndex, NodeIndex);
+type Edge = (NodeIndex, NodeIndex);
+
+struct TokenSwapper<G: GraphBase>
+where
+    G::NodeId: Eq + Hash,
+{
+    // The input graph
+    graph: G,
+    // The user-supplied mapping to use for swapping tokens
+    mapping: HashMap<G::NodeId, G::NodeId>,
+    // Number of trials
+    trials: usize,
+    // Seed for random selection of a node for a trial
+    seed: Option<u64>,
+    // Threshold for how many nodes will trigger parallel iterator
+    parallel_threshold: usize,
+    // Map of NodeId to NodeIndex
+    node_map: HashMap<G::NodeId, NodeIndex>,
+    // Map of NodeIndex to NodeId
+    rev_node_map: HashMap<NodeIndex, G::NodeId>,
+}
+
+impl<G> TokenSwapper<G>
+where
+    G: NodeCount
+        + EdgeCount
+        + IntoEdges
+        + Visitable
+        + NodeIndexable
+        + IntoNeighborsDirected
+        + IntoNodeIdentifiers
+        + Send
+        + Sync,
+    G::NodeId: Hash + Eq + Send + Sync,
+{
+    fn new(
+        graph: G,
+        mapping: HashMap<G::NodeId, G::NodeId>,
+        trials: Option<usize>,
+        seed: Option<u64>,
+        parallel_threshold: Option<usize>,
+    ) -> Self {
+        TokenSwapper {
+            graph,
+            mapping,
+            trials: trials.unwrap_or(4),
+            seed,
+            parallel_threshold: parallel_threshold.unwrap_or(50),
+            node_map: HashMap::with_capacity(graph.node_count()),
+            rev_node_map: HashMap::with_capacity(graph.node_count()),
+        }
+    }
+
+    fn map(&mut self) -> Vec<Swap> {
+        let num_nodes = self.graph.node_bound();
+        let num_edges = self.graph.edge_count();
+
+        // Directed graph with nodes matching ``graph`` and
+        // edges for neighbors closer than nodes
+        let mut digraph = StableGraph::with_capacity(num_nodes, num_edges);
+
+        // First fill the digraph with nodes. Then since it's a stable graph,
+        // must go through and remove nodes that were removed in original graph
+        for _ in 0..self.graph.node_bound() {
+            digraph.add_node(());
+        }
+        let mut count: usize = 0;
+        for gnode in self.graph.node_identifiers() {
+            let gidx = self.graph.to_index(gnode);
+            if gidx != count {
+                for idx in count..gidx {
+                    digraph.remove_node(NodeIndex::new(idx));
+                }
+                count = gidx;
+            }
+            count += 1;
+        }
```

Comment:
Instead of doing this to reproduce the holes exactly what if we maintained a mapping so that we can convert `digraph` indices to self.graph indices for output. If we changed the type of `digraph` to `StableGraph<G::NodeId, ()>` then we can store the original index in the graph when we add nodes. Something like:

```rust
for gnode in self.graph.node_identifiers {
    digraph.add_node(gnode)
}
```

Then we can just lookup the original index by looking at the weight of the node.

Reply:
Sure, that's reasonable let's take a look in a follow up. I agree it's not really a runtime concern so much. This method really just adds a single node iteration on top of everything which isn't terrible. I was thinking more about memory efficiency than runtime. BUt we can do this in a follow up.